### PR TITLE
Rich text editor component ( markdown + mentions + fileuploads)

### PR DESCRIPTION
**Pull request by UdaySagar-Git**
## Proposed Changes

This pr is a follow up of #8016 

- Rich text editor with support of bold , italic , heading styles , quote , add link , undo , redo options  
- Adds a package `turndown` to convert html to markdown 
- Adds support for mentioning users
- Adds support to upload files to specifc note

created a temporary /test route which will be removed after review 

- 
@coronasafe/care-fe-code-reviewers @coronasafe/code-reviewers

### CODE REVIEW

**Comment by UdaySagar-Git** [comment]
@gigincg @khavinshankar built a text editor from scratch. If this is okay, I will improve the styling and integrate it with doctor's notes.
There are some bugs, I will fix them along the way

https://deploy-preview-8070--care-net.netlify.app/test

File: package.json [hunk 1]
```diff
@@ -72,11 +72,13 @@
...
     "eslint-mdx": "^3.1.5",
     "events": "^3.3.0",
-    "hi-profiles": "^1.0.6",
     "glob": "^10.3.15",
+    "hi-profiles": "^1.0.6",
+    "i": "^0.3.7",
     "i18next": "^23.11.4",
     "i18next-browser-languagedetector": "^7.2.1",
     "lodash-es": "^4.17.21",
+    "npm": "^10.8.1",
```
Comment (khavinshankar) [inline package.json]: Are these packages relevant? If they were added by mistake, remove them and update the package-lock file too

File: src/Components/Common/MarkdownEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+  FaBold,
+  FaItalic,
+  FaListOl,
+  FaListUl,
+  FaUndo,
+  FaRedo,
+  FaLink,
+  FaUnlink,
+  FaQuoteRight,
+} from "react-icons/fa";
```
Comment (khavinshankar) [inline src/Components/Common/MarkdownEditor.tsx]: use care icons instead of react icons

File: src/Components/Common/MarkdownEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+  const editorRef = useRef<HTMLDivElement>(null);
+  const undoStack = useRef<string[]>([]);
+  const redoStack = useRef<string[]>([]);
+  const [isBoldActive, setIsBoldActive] = useState<boolean>(false);
+  const [isItalicActive, setIsItalicActive] = useState<boolean>(false);
+  const [isQuoteActive, setIsQuoteActive] = useState<boolean>(false);
+  const [isUnorderedListActive, setIsUnorderedListActive] =
+    useState<boolean>(false);
+  const [isOrderedListActive, setIsOrderedListActive] =
+    useState<boolean>(false);
```
Comment (khavinshankar) [inline src/Components/Common/MarkdownEditor.tsx]: consider using useReducer

File: src/Components/Common/MarkdownEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+
+  useEffect(() => {
+    document.addEventListener("selectionchange", handleSelectionChange);
+    return () => {
+      document.removeEventListener("selectionchange", handleSelectionChange);
```
Comment (khavinshankar) [inline src/Components/Common/MarkdownEditor.tsx]: Currently, event listeners are added on the document, consider adding them to editor

File: src/Components/Common/MarkdownEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+            if (listNode.parentNode) {
+              listNode.removeChild(container);
+            }
+          }
+        }
+      } else {
+        if (isUnorderedListActive || isOrderedListActive) {
+          document.execCommand("insertHTML", false, "<br><br>");
+        } else {
+          document.execCommand("insertLineBreak");
```
Comment (khavinshankar) [inline src/Components/Common/MarkdownEditor.tsx]: `document.execCommand( ... )` is deprecated, consider other alternatives

File: src/Components/Common/MarkdownEditor.tsx [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+      saveState();
+    }
+  };
+
+  const updatePreview = () => {
+    const turndownService = new TurndownService();
+    const htmlContent = editorRef.current?.innerHTML || "";
+    const markdownText = turndownService.turndown(htmlContent);
+    setMarkdown(markdownText);
+    setHtmlCode(htmlContent);
```
Comment (khavinshankar) [inline src/Components/Common/MarkdownEditor.tsx]: Currently, we are styling in HTML, converting it into markdown and then again we are converting it back into HTML. Why not directly edit markdown? Look into the feasibility of doing this.

File: src/Components/Common/MarkdownEditor.tsx [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+
+  const handleSelectionChange = () => {
+    const selection = window.getSelection();
+    if (!selection || !selection.rangeCount) return;
+
+    setIsBoldActive(document.queryCommandState("bold"));
+    setIsItalicActive(document.queryCommandState("italic"));
+    setIsQuoteActive(
+      selection.focusNode?.parentElement?.tagName === "BLOCKQUOTE",
+    );
```
Comment (khavinshankar) [inline src/Components/Common/MarkdownEditor.tsx]: There seem to be minor bugs while enabling styles (like bold, ...) without selecting the text.

**Review by khavinshankar — CHANGES_REQUESTED** [review]
@UdaySagar-Git Nicely done, the code is well broken down into functions and well written. Here are some pointers to be noted:

File: package.json (hunk 1 above)
Comment (UdaySagar-Git) [reply package.json]: i think i typed `npm i` twice for installing package 😅
will remove them

File: src/Components/Common/MarkdownEditor.tsx (hunk 2 above)
Comment (UdaySagar-Git) [reply src/Components/Common/MarkdownEditor.tsx]: sure

File: src/Components/Common/MarkdownEditor.tsx (hunk 3 above)
Comment (UdaySagar-Git) [reply src/Components/Common/MarkdownEditor.tsx]: got it1

File: src/Components/Common/MarkdownEditor.tsx (hunk 5 above)
Comment (UdaySagar-Git) [reply src/Components/Common/MarkdownEditor.tsx]: working on it !

File: src/Components/Common/MarkdownEditor.tsx (hunk 6 above)
Comment (UdaySagar-Git) [reply src/Components/Common/MarkdownEditor.tsx]: hmm.. if there is no requirement of live preview (for example showing stars * *) for bold while editing  similar to github comments , then i will go for directly editing the markdown and preview as html

File: src/Components/Common/MarkdownEditor.tsx (hunk 7 above)
Comment (UdaySagar-Git) [reply src/Components/Common/MarkdownEditor.tsx]: will look into it

**Comment by UdaySagar-Git** [comment]
Closing as superseded by #8016

**Pull request closed without merging**